features/370-auth-ui/design.md

```diff
@@ -0,0 +1,393 @@
+      * [Customizing with own templates](#customizing-with-own-templates)
+* [The authorize endpoint](#the-authorize-endpoint)
+   * [The authorization page](#the-authorization-page)
+   * [The signup page](#the-signup-page)
+   * [The signup password page](#the-signup-password-page)
+* [The settings endpoint](#the-settings-endpoint)
+* [The verification endpoint](#the-verification-endpoint)
+* [The logout endpoint](#the-logout-endpoint)
+* [Omitted features](#omitted-features)
+* [Future improvements](#future-improvements)
+   * [Customizing the signup flow or the authentication flow](#customizing-the-signup-flow-or-the-authentication-flow)
+* [Support Matrix](#support-matrix)
+* [Being a OpenID Provider](#being-a-openid-provider)
+   * [Authorization Endpoint](#authorization-endpoint)
+   * [Token Endpoint](#token-endpoint)
+   * [Token Response](#token-response)
+* [Security considerations](#security-considerations)
+   * [Being embedded in an HTML iframe](#being-embedded-in-an-html-iframe)
+* [Appendix](#appendix)
+
+## Background
+
+Auth UI is the builtin web frontend of Auth Gear.
+It provides access to most Auth Gear features as a server side rendered web application.
+It is part of Auth Gear.
+
+It is intended to be a conforming OpenID Provider implementation of Auth Gear in the future.
+As of the time of writing, the conformance of OpenID Provider is not a goal.
+However, the implementation must be prepared for conformance in the future.
+
+## Where Auth UI is served
+
+It is served in the same domain as Auth Gear and Microservices.
+
+### Reason 1: Individual gears should be unaware of Custom Domain
+
+Auth UI is going to be responsible for user verification. The verification email includes an URL to Auth UI. This implies Auth Gear has to know where Auth UI is served. When Custom Domain was introduced, it is a feature solely handled by Gateway. Individual gears such as Auth Gear are unaware of Custom Domain. Serving Auth UI in a different domain will break this property. If Auth UI is served in the same domain, it is trivial for Auth Gear to know where it is.
+
+### Reason 2: Cookies written by Auth Gear is host-only
+
+A host-only cookie only matches the domain by simple string comparison. While a non host-only cookie matches the domain via the [domain-matches](https://tools.ietf.org/html/rfc6265#section-5.1.3) algorithm. To allow Auth UI to be embedded within an iframe in a Skygear App, the session and its cookie must be shared by both. This means the domain has to be either exact match or domain-matches.
+
+If cookies written by Auth Gear have to be not host-only, then the cookie attribute `Domain` must be specified. If the custom domain is `www.example.com`, then the value of the attribute should be the apex domain `example.com`. So Auth Gear cannot avoid being aware of Custom Domain.
+
+### The actual location of Auth UI
+
+Auth UI is served at the path prefix `/_auth/_/`
```

Comment:
So can we support, as we said, having something like `auth.oursky.com/authorize` when custom domain name need to be supported?

My understanding here is not planned -- but I think this is a required features. The main reason is for user experience, the URL have to be easily recognizable to avoid phishing problem and avoid trust problem with users.

Reply:
and I think it kind of implies AuthUI can't live under `/_auth/`

Reply:
So in consideration with the cookies thing, I'm thinking is it better if we...

A. force certain rules on the domain, if skygear custom domain is `example.com`, then auth UI custom domain must be `*.example.com`; OR
B. we go directly make Auth UI totally independent, so after AuthUI authenticated the users, it returns a token to call back Auth Gear to generate the real session.

Reply:
I have the following proposal:

1. Auth UI is deployed standalone. It shares the same code base of Auth Gear.
1. Without custom domain Auth UI is served at `https://accounts.<appname>.skygearapp.com`. The choice of `accounts` is borrowed from Google. Later on when we have deployment version, Auth UI has higher precedence. In practice it should not be a problem.
1. Add a new custom domain feature to allow configure a custom domain for Auth UI.
1. Auth UI and Auth Gear somehow can detect if the request domain is a custom domain. If it is a custom domain, use public suffix list to infer the eTLD+1 and write the cookie with `Domain` attribute. The cookies for custom domain are no longer host only and are shared by **all** subdomains including those unknown to Skygear Controller. The cookie name Auth Gear and Auth UI use for storing the access token is called `session`. It may conflict with other systems running other unknown subdomains.
1. If the request domain is not a custom domain, then the domain of the cookie is derived by removing the leftmost label. 
1. `Content-Security-Policy` default value is still `frame-ancestors 'self'`. It means must explicitly set it to something else if they want to use Auth UI in an iframe.

One downside is that it is impossible to make Auth UI share the domain with other gears and microservices.

cc @kiootic